README.md
# umapscan

[![CRAN status](https://www.r-pkg.org/badges/version-ago/umapscan)](https://cran.r-project.org/package=umapscan)
[![Lifecycle: experimental](https://img.shields.io/badge/lifecycle-experimental-orange.svg)](https://www.tidyverse.org/lifecycle/#experimental)
[![R build status](https://github.com/juba/umapscan/workflows/R-CMD-check/badge.svg)](https://github.com/juba/umapscan/actions?query=workflow%3AR-CMD-check)

This is an helper package for using UMAP + DBSCAN in R.

This package can't / must not be used right now :

- highly experimental : everything is subject to breaking changes
- no documentation
- bugs, bugs, bugs

You've been warned !



## TODO

- noise_inherit_parent should return <Noise> if no no-noise parent


NEWS.md
# umapscan 0.0.1

* Initial version


tests/testthat.R
library(testthat)
library(umapscan)

test_check("umapscan")


tests/testthat/test_plots.R
context("umapscan plots")

skip_on_cran()
skip_on_travis()
skip_on_ci()

iris_num <- iris[, c("Sepal.Length", "Sepal.Width", "Petal.Length", "Petal.Width")]
iris_sup <- iris[, "Species", drop=FALSE]
us <- new_umapscan(iris_num, data_sup = iris_sup, seed = 1337, scale = FALSE)

us_plot <- plot(us)
us_plot_color <- plot(us, color = Species)
us_plot_full <- plot(us, color = Species, label = "test", ellipses = TRUE, alpha = 0.8, fixed = TRUE)

vdiffr::expect_doppelganger("Base umapscan plot", us_plot)
vdiffr::expect_doppelganger("Colored umapscan plot", us_plot_color)
vdiffr::expect_doppelganger("Full umapscan plot", us_plot_full)


tests/testthat/test_umapscan.R
context("umapscan class functions")

iris_num <- iris[, c("Sepal.Length", "Sepal.Width", "Petal.Length", "Petal.Width")]
iris_sup <- iris[, "Species", drop=FALSE]

test_that("new_umapscan with wrong arguments throws error", {
  expect_error(new_umapscan(d = 1:10))
  expect_error(new_umapscan(d = iris_num, data_sup = 1:10))
  expect_error(new_umapscan(d = iris_num, data_sup = iris_sup[-1, ]))
})

test_that("new_umapscan results are ok", {
  us <- new_umapscan(iris_num, data_sup = iris_sup, seed = 1337, scale = FALSE)
  expect_equal(us$data, iris_num)
  expect_equal(us$data_sup, iris_sup)
  expect_equal(dim(us$cluster), c(0, 5))
  expect_equal(dim(us$umap), c(nrow(iris_num), 2))
  set.seed(1337)
  umap <- uwot::umap(iris_num, n_sgd_threads = 0)
  expect_equal(tibble::tibble(x = umap[,1], y = umap[,2]), us$umap)
})

test_that("new_umapscan results with scale=TRUE are ok", {
  us <- new_umapscan(iris_num, data_sup = iris_sup, seed = 24312, scale = TRUE)
  set.seed(24312)
  umap <- uwot::umap(dplyr::mutate_all(iris_num, base::scale),  n_sgd_threads = 0)
  expect_equal(tibble::tibble(x = umap[,1], y = umap[,2]), us$umap)
})

test_that("new_umapscan results are reproducible", {
  us <- new_umapscan(iris_num, seed = 24312, scale = TRUE)
  # saveRDS(us$umap %>% dplyr::slice(1:50) %>% data.frame, "tests/values/umapscan1.rds")
  expect_equal(
    us$umap %>% dplyr::slice(1:50) %>% data.frame,
    readRDS("../values/umapscan1.rds")
  )
})

test_that("set.seed is ok", {
  set.seed(556677)
  x <- runif(1)
  expect_equal(x, 0.05051134, tolerance = 0.00000001)
  y <- rnorm(1)
  expect_equal(y, 1.970592, tolerance = 0.000001)
  z <- runif(1)
  expect_equal(z, 0.3479608, tolerance = 0.0000001)
})

test_that("base umapscan with seed is ok", {
  set.seed(8222)
  umap <- uwot::umap(USArrests, n_sgd_threads = 0)
  # saveRDS(umap, "tests/values/umapscan2.rds")
  expect_equivalent(
    umap,
    readRDS("../values/umapscan2.rds")
  )
  set.seed(8223)
  umap_pca <- uwot::umap(USArrests, init = "pca", n_sgd_threads = 0)
  # saveRDS(umap_pca, "tests/values/umapscan3.rds")
  expect_equivalent(
    umap_pca,
    readRDS("../values/umapscan3.rds")
  )
})

test_that("uwot version is the same", {
  uwot_version <- installed.packages()["uwot",]
  res <- c(Package = "uwot",
    Version = "0.1.5", Priority = NA_character_, Depends = "Matrix", Imports = "Rcpp, methods, FNN, RSpectra, RcppAnnoy (>= 0.0.11),\nRcppParallel, irlba",
    LinkingTo = "Rcpp, RcppProgress, RcppParallel, RcppAnnoy, dqrng",
    Suggests = "testthat, covr", Enhances = NA_character_, License = "GPL-3",
    License_is_FOSS = NA_character_, License_restricts_use = NA_character_, OS_type = NA_character_,
    MD5sum = NA_character_, NeedsCompilation = "yes", Built = "3.6.2")
  expect_equal(uwot_version[names(uwot_version) != "LibPath"], res)
  })




R/utils.R
#' Return a qualitative color palette depending on the number of values
#' of its `var` argument
#'
#' @param var vector to compute palette values from
#' @param label title of generated paletee
#' @param type type of color palette
#'
#' @importFrom ggsci scale_color_d3 scale_fill_d3

qualitative_palette <- function(var, label, type = c("color", "fill")) {

  type <- match.arg(type)

  var <- var[!is.na(var)]
  n_values <- dplyr::n_distinct(var)

  if (n_values <= 9) {
    if (type == "color") {
      return(scale_color_brewer(label, palette = "Set1", na.value = "grey50"))
    }
    if (type == "fill") {
      return(scale_fill_brewer(label, palette = "Set1", na.value = "grey50"))
    }

  }

  if (n_values <= 12) {
    if (type == "color") {
      return(scale_color_brewer(label, palette = "Paired", na.value = "grey50"))
    }
    if (type == "fill") {
      return(scale_fill_brewer(label, palette = "Paired", na.value = "grey50"))
    }
  }

  ## n_values > 12
  if (type == "color") {
    return(ggsci::scale_color_d3(palette = "category20", name = label, na.value = "white"))
  }
  if (type == "fill") {
    return(ggsci::scale_fill_d3(palette = "category20", name = label, na.value = "white"))
  }


}


R/umapscan.R
#' Compute UMAP embeddings from a numerical data frame
#'
#' @param d a data frame of numeric variables
#' @param n_neighbors `n_neighbors` argument passed to [uwot::umap()]
#' @param min_dist `min_dist` argument passed to [uwot::umap()]
#' @param metric `metric` argument passed to [uwot::umap()]
#' @param scale scaling to be applied to the variables. Value passed to [uwot::umap()].
#' @param seed integer seed passed to [base::set.seed()] before umap embeddings for
#'   reproductibility
#' @param data_sup a data.frame of supplementary variables not used for UMAP
#'   embeddings but for results visualisation
#' @param ... other arguments passed to [uwot::umap()]
#'
#' @return
#' A list of class `umapscan` with the following components :
#' - `umap` : results of UMAP embeddings
#' - `data` : original numerical dataset
#' - `data_sup` : supplementary dataset
#' - `clusters` : a tree-like hierarchy of clusters
#' - `call` : function call
#' - `seed` : the seed used for random numbers for all umap and dbscan operations
#'
#' @seealso [uwot::umap()]
#'
#' @export
#' @importFrom uwot umap
#' @importFrom tibble tibble
#'
#' @examples
#' library(dplyr)
#' iris_num <- iris %>% select_if(is.numeric)
#' iris_sup <- iris %>% select(Species)
#' new_umapscan(iris_num, data_sup = iris_sup)


new_umapscan <- function(
  d,
  n_neighbors = 15,
  min_dist = 0.01,
  metric = "euclidean",
  scale = FALSE,
  seed,
  data_sup = NULL,
  ...
) {

  if (!inherits(d, "data.frame")) stop("d must be a numerical data frame.")
  if (!is.null(data_sup)) {
    if (!inherits(data_sup, "data.frame")) stop("data_sup must be a data frame.")
    if (nrow(data_sup) != nrow(d)) stop("d and data_sup must have the same number of rows.")
  }

  d_comp <- d

  if (missing(seed)) { seed <- round(stats::runif(1) * 100000) }

  set.seed(seed, kind = "default", normal.kind = "default", sample.kind = "default")
  umap <- uwot::umap(
    d_comp,
    n_neighbors = n_neighbors,
    min_dist = min_dist,
    metric = metric,
    scale = scale,
    n_sgd_threads = 0,
    ...
  )
  umap <- tibble::tibble(
    x = umap[,1],
    y = umap[,2],
  )

  clusters <- tibble::tibble(
    from = character(0),
    to = character(0),
    n = integer(0),
    ids = list(),
    level = integer(0)
  )

  res <- list(
    umap = umap,
    data = d,
    clusters = clusters,
    data_sup = data_sup,
    call = match.call(),
    seed = seed
  )

  class(res) <- c(class(res), "umapscan")

  res
}


#' Print an umapscan object
#'
#' @param x umapscan object to be printed
#' @param ... arguments passed to other methods
#'
#' @export
#' @importFrom data.tree as.Node

print.umapscan <- function(x, ...) {

  cat("\nCall: ", paste(deparse(x$call), sep = "\n", collapse = "\n"),
    "\n\n", sep = "")
  dims <- dim(x$data)
  cat("UMAP embeddings of a ", dims[1], "x", dims[2], " data frame\n", sep = "")

  if (!is.null(x$data_sup)) {
    dims_sup <- dim(x$data_sup)
    cat ("with a ", dims_sup[1], "x", dims_sup[2],
      " data frame of supplementary data\n", sep = "")
  }

  cat("\nClusters :")
  if (nrow(x$clusters) > 0) {
    cat("\n\n")
    print(data.tree::as.Node(x$clusters, mode = "network"))
  } else {
    cat(" <none>\n\n")
  }

  invisible(x)
}



#' Plot an umapscan object
#'
#' @param x umapscan object to be plotted
#' @param color variable to highlight points (`tidylang` style). Must be a variable
#'   of the `data` or `data_sup` elements of `us`
#' @param palette `viridis` (for continuous `color`) or `brewer` (for categorical
#'   `color`) palette name
#' @param label legend label
#' @param ellipses if TRUE and `color` is categorical, add repartition ellipses
#' @param alpha points transparency
#' @param fixed if TRUE, add [ggplot2::coord_fixed()] to plot
#' @param ... arguments passed to other methods
#'
#' @examples
#' library(dplyr)
#' iris_num <- iris %>% select_if(is.numeric)
#' iris_sup <- iris %>% select(Species)
#' us <- new_umapscan(iris_num, data_sup = iris_sup, n_neighbors = 25, min_dist = 0.1)
#' plot(us, color = Species, ellipses = TRUE)
#' plot(us, color = Sepal.Width, palette = "viridis", alpha = 0.5, fixed = TRUE)
#'
#' @export
#' @import ggplot2
#' @import dplyr

plot.umapscan <- function(
  x, color, palette, label = NULL, ellipses = FALSE, alpha = 0.2, fixed = FALSE, ...
) {

  if (!inherits(x, "umapscan")) stop("`x` must be an object of class umapscan.")

  d <- dplyr::bind_cols(x$umap, x$data, x$data_sup)

  g <- ggplot(d, aes(x = .data$x, y = .data$y)) +
    xlab("") +
    ylab("")

  if (fixed) {
    g <- g + coord_fixed()
  }

  if (missing(color)) {
    g <- g + geom_point(alpha = alpha)
    return(g)
  }

  g <- g + geom_point(aes(color = {{color}}), alpha = alpha)

  col_var <- d %>% dplyr::pull({{color}})

  if (is.numeric(col_var)) {
    if (missing(palette)) palette <- "inferno"
    g <- g + scale_color_viridis_c(label, option = palette)
  } else {
    if (missing(palette)) {
      scale <- qualitative_palette(col_var, label)
    } else {
      scale <- scale_color_brewer(label, palette = palette)
    }
    g <- g + scale
  }

  if (ellipses) {
    g <- g + stat_ellipse(aes(color = {{color}}))
  }

  g

}



#' Display an umapscan clustering as a leaflet map.
#'
#' @param us umapscan object to be plotted
#' @param point_labels a character vector of point labels displayed
#'   in a popup on hover. Can contain HTML.
#'
#' @export
#' @importFrom purrr map
#' @importFrom shiny HTML
#' @import leaflet

map_plot <- function(us, point_labels = NULL) {

  ## Generate data for each clustering level
  levels <- purrr::map(1:max(us$clusters$level), function(level) {

    clusters <- get_clusters_membership(us, max_level = level)
    clusters[clusters == "<Noise>"] <- NA
    suppressWarnings({
      pal <- leaflet::colorFactor(
        palette = "Paired", domain = clusters, na.color = "white"
      )
    })

    clusters_labels <- us$umap %>%
      mutate(clust = clusters) %>%
      tidyr::drop_na(.data$clust) %>%
      group_by(.data$clust) %>%
      summarise(x = mean(.data$x), y = mean(.data$y))

    list(clusters = clusters, palette = pal, clusters_labels = clusters_labels)
  })

  ## Map object
  map <- leaflet::leaflet(
    us$umap,
    options = leaflet::leafletOptions(preferCanvas = TRUE)
  )

  ## Data point labels
  if (!is.null(point_labels)) {
    point_labels <- purrr::map(point_labels, shiny::HTML)
  }

  # Add group layers
  current_group <- 0
  for (level in levels) {
    current_group <- current_group + 1
    group <- paste("Level", current_group)
    suppressWarnings({
      map <- map %>%
        ## Add data points
        leaflet::addCircleMarkers(
          lng = ~x, lat = ~y,
          stroke = FALSE,
          radius = 5,
          fillColor = level$palette(level$clusters),
          fillOpacity = 0.5,
          label = point_labels,
          group = group
        ) %>%
        ## Add cluster points
        leaflet::addCircleMarkers(
          data = level$clusters_labels,
          lng = ~x, lat = ~y,
          label = ~clust,
          labelOptions = leaflet::labelOptions(permanent = TRUE, direction = "top"),
          fillColor = ~level$palette(clust),
          fillOpacity = 0.9,
          radius = 10,
          weight = 2,
          opacity = 1,
          color = "white",
          group = group
        ) %>%
        leaflet::hideGroup(group)
    })
  }

  map <- map %>%
    leaflet::showGroup("Level 1") %>%
    leaflet::addLayersControl(
      baseGroups = paste("Level", seq_along(levels)),
      options = leaflet::layersControlOptions(collapsed = FALSE)
    )

  map

}








R/clustering.R
#' Compute DBSCAN clusters from a umapscan object
#'
#' This function runs DBSCAN with the specified arguments to compute new clusters
#' on a umapscan object. Only points which don't have an already validated cluster
#' are taken into account.
#'
#' @param us umapscan object
#' @param parent name of the parent cluster
#' @param noise_only only compute clusters for current 'Noise' points
#' @param eps `eps` argument passed to [dbscan::dbscan()]
#' @param minPts `minPts` argument passed to [dbscan::dbscan()]
#' @param graph if TRUE, display a plot of the computed clusters
#' @param alpha point transparency for clusters plot
#' @param ellipses if TRUE, plot confidence ellipses around clusters
#'
#' @return
#' Returns an updated `umapscan` object, and optionally displays a clusters plot.
#'
#' @seealso
#' [new_umapscan()], [describe_clusters()], [get_cluster_data()], [rename_cluster()],
#' [plot_clusters()], [get_clusters_membership()]
#'
#' @export
#'
#' @examples
#' library(dplyr)
#' iris_num <- iris %>% select_if(is.numeric)
#' us <- new_umapscan(iris_num, n_neighbors = 25, min_dist = 0.1, seed = 1337)
#' us <- compute_clusters(us, minPts = 3, eps = 0.5)
#' compute_clusters(us, minPts = 3, eps = 0.45, alpha = 1, parent = "3")
#'
#' @importFrom dbscan dbscan hdbscan

compute_clusters <- function(us, parent = "", noise_only = FALSE, eps, minPts, graph = TRUE, alpha = 1, ellipses = TRUE) {

  if (!inherits(us, "umapscan")) stop("`us` must be an object of class umapscan.")

  if (noise_only) {
    ids <- get_noise_ids(us, parent)
  } else {
    ids <- get_ids(us, parent)
  }
  us <- remove_noise_cluster(us, parent)

  d_clust <- us$umap %>% slice(ids)
  set.seed(us$seed, kind = "default", normal.kind = "default", sample.kind = "default")
  if (missing(eps)) {
    message("Missing eps, hdbscan performed instead of dbscan")
    db <- dbscan::hdbscan(d_clust, minPts = minPts)
  } else {
    db <- dbscan::dbscan(d_clust, eps = eps, minPts = minPts)
  }

  clust <- db$cluster
  parent_string <- ifelse(parent == "", "", paste0(parent, "_"))
  clust <- paste0(parent_string, clust)
  clust[db$cluster == 0] <- "<Noise>"
  d_clust$cluster <- clust

  if (parent == "") {
    level <- 1
  } else {
    level <- us$clusters$level[us$clusters$to == parent] + 1
  }

  for (cl in unique(clust)) {
    if (is.na(cl)) next
    select <- (clust == cl) & !is.na(clust)
    line <- tibble::tibble(
      from = parent,
      to = cl,
      n = length(ids[select]),
      ids = list(ids[select]),
      level = level
    )
    us$clusters <- dplyr::bind_rows(us$clusters, line)
  }

  select <- us$clusters$to != "<Noise>"
  us$clusters$to[select] <- make.unique(us$clusters$to[select])

  if (graph) {
    g <- plot_clusters(us, parent, alpha = alpha, fixed = TRUE)
    print(g)
  }

  us
}



#' Plot clusters of an umapscan object
#'
#' @param us umapscan object
#' @param parent name of the parent cluster
#' @param noise_inherit_parent if TRUE, 'Noise' points are given their parent cluster
#'   membership
#' @param alpha point transparency for clusters plot
#' @param ellipses if TRUE, plot confidence ellipses around clusters
#' @param fixed if TRUE, force coord_fixed on plot
#'
#' @seealso
#' [compute_clusters()], [describe_clusters()]
#' @export
#'
#' @examples
#' library(dplyr)
#' iris_num <- iris %>% select_if(is.numeric)
#' us <- new_umapscan(iris_num, n_neighbors = 25, min_dist = 0.1, seed = 1337)
#' us <- compute_clusters(us, minPts = 3, eps = 0.5, graph = FALSE)
#' plot_clusters(us, alpha = 1)
#' us <- compute_clusters(us, minPts = 3, eps = 0.45, graph = FALSE, parent = "3")
#' plot_clusters(us, alpha = 0.5, ellipses = FALSE, parent = "3")
#' plot_clusters(us)
#'
#' @import ggplot2
#' @importFrom tidyr drop_na

plot_clusters <- function(us, parent = "", noise_inherit_parent = FALSE, alpha = 1, ellipses = TRUE, fixed = FALSE) {

  clust <- get_clusters_membership(us, parent, noise_inherit_parent = noise_inherit_parent)
  if (all(is.na(clust))) stop("No defined clusters in umapscan object.")
  d_clust <- us$umap[!is.na(clust),]
  clust <- clust[!is.na(clust)]
  clust[clust == "<Noise>"] <- NA
  d_clust$cluster <- clust
  color_scale <- qualitative_palette(clust, label = "Cluster")

  g <- ggplot(d_clust, aes(x = .data$x, y = .data$y, color = factor(.data$cluster))) +
    geom_point(size = 1, alpha = alpha) +
    color_scale +
    guides(colour = guide_legend(override.aes = list(alpha = 1))) +
    xlab("") +
    ylab("")

  if (ellipses) {
    d_ellipses <- d_clust %>% tidyr::drop_na(.data$cluster)
    g <- g + stat_ellipse(data = d_ellipses, na.rm = TRUE)
  }

  if (fixed) {
    g <- g + coord_fixed()
  }

  g
}


#' Describe clusters of an umapscan object
#'
#' @param us umapscan object to describe clusters
#' @param parent name of the parent cluster
#' @param type plot type, either `"boxplot"` or `"ridges"`
#'
#' @seealso
#' [compute_clusters()], [get_cluster_data()], [get_clusters_membership()],
#' [rename_cluster()]
#'
#' @export
#' @importFrom tidyr pivot_longer
#' @importFrom ggridges geom_density_ridges
#' @import dplyr
#'
#' @examples
#' library(dplyr)
#' iris_num <- iris %>% select_if(is.numeric)
#' us <- new_umapscan(iris_num, n_neighbors = 25, min_dist = 0.1, seed = 1337)
#' us <- compute_clusters(us, minPts = 3, eps = 0.5)
#' describe_clusters(us)
#' us <- compute_clusters(us, minPts = 3, eps = 0.45, parent = "3")
#' describe_clusters(us, type = "ridge")
#' describe_clusters(us, parent = "3")

describe_clusters <- function(us, parent = "", type = c("boxplot", "ridges")) {

  type <- match.arg(type)

  clusters <- get_clusters_membership(us, parent, noise_inherit_parent = FALSE)
  clusters[clusters == "<Noise>"] <- NA

  select <- !is.na(clusters)
  d <- us$data %>% dplyr::filter(select)
  d$cluster <- clusters[select]

  d_long <- d %>%
    tidyr::pivot_longer(-.data$cluster) %>%
    dplyr::mutate(
      cluster = factor(.data$cluster)
    )

  fill_scale <- qualitative_palette(d_long$cluster, "Cluster", type = "fill")

  if (type == "boxplot") {
    g <- ggplot(d_long, aes(x = .data$cluster, y = .data$value, fill = .data$cluster)) +
      geom_boxplot(outlier.shape = NA) +
      fill_scale +
      facet_wrap(~name, scales = "free_y") +
      theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
      xlab("") +
      ylab("")
  }

  if (type == "ridges") {

    g <- ggplot(d_long, aes(x = .data$value, y = .data$cluster, fill = .data$cluster)) +
      ggridges::geom_density_ridges(alpha = .7) +
      fill_scale +
      facet_wrap(~name, scales = "free_x") +
      guides(fill = guide_legend(override.aes = list(alpha = 1)))
  }

  g

}


#' Get cluster membership for each observation of an umapscan object
#'
#' The clusters taken into account are all the "leaves" of the object
#' clusters tree. If a `parent` is specified, only the leaves of its
#' subtree are taken.
#'
#' @param us umapscan object to describe clusters
#' @param parent name of the parent cluster
#' @param max_level get membership at most this level deep
#' @param noise_inherit_parent if TRUE, 'Noise' points are given their parent cluster
#'   membership
#'
#' @seealso
#' [compute_clusters()], [get_cluster_data()], [rename_cluster()]
#'
#' @export
#'
#' @examples
#' library(dplyr)
#' iris_num <- iris %>% select_if(is.numeric)
#' us <- new_umapscan(iris_num, n_neighbors = 25, min_dist = 0.1, seed = 1337)
#' us <- compute_clusters(us, minPts = 3, eps = 0.5)
#' get_clusters_membership(us)
#' us <- compute_clusters(us, minPts = 3, eps = 0.45, parent = "3")
#' get_clusters_membership(us)
#' get_clusters_membership(us, parent = "3")


get_clusters_membership <- function(us, parent = "", max_level, noise_inherit_parent = FALSE) {

  if (parent == "<Noise>") {
    stop("Can't get membership starting from a <Noise> node")
  }

  if (missing(max_level)) {
    tree <- us$clusters
  } else {
    tree <- us$clusters %>%
      filter(.data$level <= max_level)
  }

  clusters <- rep(NA, nrow(us$data))

  if (nrow(us$clusters) == 0) {
    return(clusters)
  }

  leaves <- get_leaves(tree, node = parent)

  for (i in 1:nrow(leaves)) {
    leaf <- leaves[i,]
    if (leaf$to == "<Noise>") {
      if (noise_inherit_parent) {
        clusters[get_noise_ids(us, leaf$from)] <- leaf$from
      } else {
        clusters[get_noise_ids(us, leaf$from)] <- leaf$to
      }
    } else {
      clusters[get_ids(us, leaf$to)] <- leaf$to
    }
  }

  clusters
}



#' Get leaves from a node
#'
#' @param tree tree object, such as a `clusters` element of a `umapscan` object
#' @param node node to find leaves from
#' @param parent optional parent of `node` (should not be used directly, only for recursive call).
#'
#' `parent` is used during computation to differentiate different 'Noise' nodes.
#'
#' @importFrom dplyr bind_rows
#' @importFrom tibble tibble

get_leaves <- function(tree, node = "", parent = NA) {

  if (node == "<Noise>" && is.na(parent)) {
    stop("Can't get leaves from a <Noise> node.")
  }

  children <- tree$to[tree$from == node]

  if (length(children) == 0) {
    return(tibble::tibble(from = parent, to = node))
  }
  leaves <- tibble::tibble(from = character(0), to = character(0))
  for(child in children) {
    leaves <- dplyr::bind_rows(leaves, get_leaves(tree, child, parent = node))
  }
  return(leaves)
}


#' Get ids from a node name
#'
#' @param us an umapscan object
#' @param node node to get ids from
#'
#' @import dplyr

get_ids <- function(us, node) {

  if (node == "") {
    return(1:nrow(us$data))
  }

  if (node == "<Noise>") {
    stop("Can't get ids from a <Noise> node.")
  }

  us$clusters %>%
    filter(.data$to == node) %>%
    pull(.data$ids) %>%
    unlist
}

#' Get ids from a 'Noise' child of a node
#'
#' @param us an umapscan object
#' @param node node to get child 'Noise' ids from
#'
#' @import dplyr

get_noise_ids <- function(us, node) {

  if (node == "<Noise>") {
    stop("Can't get ids from a <Noise> node.")
  }

  us$clusters %>%
    filter(.data$from == node, .data$to == "<Noise>") %>%
    pull(.data$ids) %>%
    unlist
}


#' Get an umapscan cluster data
#'
#' @param us umapscan object
#' @param cluster cluster to retrieve
#'
#' @seealso
#' [compute_clusters()], [describe_clusters()], [rename_cluster()]
#'
#' @return
#' 1 data frame with the `data` and `data_sup` variables of the cluster observations
#' @export
#' @import dplyr
#'
#' @examples
#' library(dplyr)
#' iris_num <- iris %>% select_if(is.numeric)
#' iris_sup <- iris %>% select(Species)
#' us <- new_umapscan(iris_num, data_sup = iris_sup, n_neighbors = 25, min_dist = 0.1, seed = 1337)
#' us <- compute_clusters(us, minPts = 3, eps = 0.5)
#' get_cluster_data(us, "1")

get_cluster_data <- function(us, cluster) {

  if (cluster == "<Noise>") {
    stop("Can't get data for a <Noise> cluster.")
  }

  ids <- get_ids(us, cluster)
  d <- dplyr::bind_cols(us$data_sup, us$data)
  d %>%
    dplyr::slice(ids) %>%
    dplyr::mutate(umapscan_cluster = cluster)

}



#' Rename an umapscan cluster
#'
#' @param us umapscan object
#' @param old current cluster identifier
#' @param new new cluster label
#'
#' @seealso
#' [compute_clusters()], [describe_clusters()]
#'
#' @return
#' An updated umapscan object (invisibly). If two clusters have the same name after
#' renaming, they are merged together.
#' @export
#' @import dplyr
#'
#' @examples
#' library(dplyr)
#' iris_num <- iris %>% select_if(is.numeric)
#' us <- new_umapscan(iris_num, n_neighbors = 25, min_dist = 0.1, seed = 1337)
#' us <- compute_clusters(us, minPts = 3, eps = 0.5)
#' rename_cluster(us, "1", "Cluster 1")
#' us

rename_cluster <- function(us, old, new) {

  if (old == "<Noise>") {
    stop("Can't rename a <Noise> cluster.")
  }

  ## Test for same new name with another parent
  if (new != "<Noise>" && new %in% us$clusters$to) {
    parent_old <- us$clusters$from[us$clusters$to == old]
    parent_new <- us$clusters$from[us$clusters$to == new]
    if (parent_old != parent_new) {
      stop("Can't rename a cluster with the same name as another cluster with another parent.")
    }
  }

  us$clusters <- us$clusters %>%
    mutate(
      from = if_else(.data$from == old, new, .data$from),
      to = if_else(.data$to == old, new, .data$to),
    ) %>%
    group_by(.data$from, .data$to, .data$level) %>%
    summarise(
      n = sum(n),
      ids = list(unlist(c(.data$ids)))
    ) %>%
    ungroup()

  us
}


#' Remove a cluster from an umapscan object
#'
#' If the cluster has children, they will be removed too.
#'
#' @param us umapscan object
#' @param cluster label of the cluster to remove
#' @param rm_root if TRUE, also remove the root cluster node. Otherwise, only remove
#'   its children (should not be used directly, only for recursive call).
#'
#' @return
#' An updated umapscan object.
#' @importFrom purrr pwalk
#'
#' @examples
#' library(dplyr)
#' iris_num <- iris %>% select_if(is.numeric)
#' us <- new_umapscan(iris_num, n_neighbors = 25, min_dist = 0.1, seed = 1337)
#' us <- compute_clusters(us, minPts = 3, eps = 0.5)
#' us <- compute_clusters(us, parent = "3" ,minPts = 3, eps = 0.45)
#' remove_cluster(us, "3_1")
#' us
#' remove_cluster(us, "3")
#' us

remove_cluster <- function(us, cluster, rm_root = FALSE) {

  if (cluster == "<Noise>") {
    stop("Can't remove a <Noise> cluster.")
  }

  from_lines <- us$clusters %>%
    filter(.data$from == cluster) %>%
    select(this_from = .data$from, this_to = .data$to)

  purrr::pwalk(from_lines, function(this_from, this_to) {
    ## Ensure we only remove one <Noise> node
    if (this_to == "<Noise>") {
      us$clusters <<- us$clusters %>%
        filter(!(.data$from == this_from & .data$to == this_to))
    } else {
      us <<- remove_cluster(us, this_to, rm_root = TRUE)
    }
  })

  if (rm_root) {
    to_lines <- us$clusters$to == cluster
    us$clusters <- us$clusters %>% filter(!to_lines)
  }

  us
}


#' Remove a 'Noise' cluster from an umapscan object
#'
#' @param us umapscan object
#' @param parent parent of the 'Noise' cluster to remove

remove_noise_cluster <- function(us, parent) {

  if (parent == "<Noise>") {
    stop("Can't remove a <Noise> cluster from a <Noise> parent.")
  }

  us$clusters <- us$clusters %>%
    filter(!(.data$from == parent & .data$to == "<Noise>"))

  return(us)
}



#' Reinit all clusters of an umapscan object
#'
#' @param us umapscan object
#'
#' @return
#' An updated umapscan object with an empty clusters element.
#'
#' @export
#' @importFrom tibble tibble

reinit_clusters <- function(us) {
  us$clusters <- tibble::tibble(
    from = character(0),
    to = character(0),
    n = integer(0),
    ids = list()
  )

  us
}


#' Collapse a clusters tree
#'
#' @param us an umapscan object
#' @param collapsed list of nodes to collapse
#'
#' @return
#' An updated umapscan object with `collapsed` nodes children removed.
#'
#' @export

collapse_clusters <- function(us, collapsed) {

  for (node in collapsed) {
    us <- remove_cluster(us, node)
  }

  us
}




